### PR TITLE
Form traits

### PR DESCRIPTION
**Pull request by guusw**

### CODE REVIEW

File: include/shards/shards.h [hunk 1]
```diff
@@ -404,6 +412,23 @@ struct SHTypeInfo {
   SHBool recursiveSelf;
 };
 
+typedef struct SHTraitVariable {
+  // Name of the variable
+  SHString name;
```
Comment (sinkingsugar) [inline include/shards/shards.h]: this should be SHStringWithLen, we should deprecate SHString usage

File: include/shards/shards.h [hunk 2]
```diff
@@ -404,6 +412,23 @@ struct SHTypeInfo {
...
+  // Expected variable type
+  struct SHTypeInfo type;
+} SHTraitVariable;
+SH_ARRAY_DECL(SHTraitVariables, struct SHTraitVariable);
+
+typedef struct SHTrait {
+  // Unique Id of the trait, which is also the hash of the items
+  uint64_t id[2];
+  // Friendly name given to this trait, unhashed
+  SHString name;
```
Comment (sinkingsugar) [inline include/shards/shards.h]: SHStringWithLen

File: include/shards/shards.h [hunk 3]
```diff
@@ -977,11 +1034,18 @@ typedef const struct SHObjectInfo *(__cdecl *SHFindObjectInfo)(int32_t vendorId,
...
+typedef void(__cdecl *SHStringGrow)(SHStringPayload *str, uint32_t newCap);
+typedef void(__cdecl *SHStringFree)(SHStringPayload *str);
+
 typedef struct _SHCore {
   // Aligned allocator
   SHAlloc alloc;
   SHFree free;
 
+  // String interface
+  SHStringGrow stringGrow;
```
Comment (sinkingsugar) [inline include/shards/shards.h:1046]: We should deprecate SHString, if this uses SHString

File: shards/core/trait.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#include <cstdlib>
+
+namespace shards {
+void freeTraitVariable(SHTraitVariable &ivar) {
+  std::free((void *)ivar.name);
+  freeDerivedInfo(ivar.type);
+}
+SHTraitVariable cloneTraitVariable(const SHTraitVariable &other) {
+  SHTraitVariable clone{
+      .name = strdup(other.name),
```
Comment (sinkingsugar) [inline shards/core/trait.cpp]: no no please use SHStringWithLen or rust goes crazy

File: shards/lang/src/eval.rs [hunk 5]
```diff
@@ -1143,31 +1227,62 @@ fn combine_namespaces(partial: &RcStrWrapper, fully_qualified: &RcStrWrapper) ->
   combined.join("/").into()
 }
 
+struct ResolvedVar {
```
Comment (sinkingsugar) [inline shards/lang/src/eval.rs]:
```
enum ResovedVar {
  Const(SVar),
  NonConst(SVar),
  }
 ```
 
 Maybe

File: shards/lang/src/shards.pest [hunk 6]
```diff
@@ -21,16 +21,16 @@ None          =  { "none" | "null" }
 ConstValue    =  { None | Boolean | VarName | Number | String | Seq | Table | Enum }
 Shards        =  { "{" ~ Sequence? ~ "}" }
 Block         = _{ Func | EvalExpr | Expr | Enum | Shard | TakeTable | TakeSeq | ConstValue | Shards }
-Iden          = @{ (ASCII_ALPHA | "_" | "*") ~ (ASCII_ALPHANUMERIC | "_" | "-" | ".")* }
+Iden          = @{ (ASCII_ALPHA | "_" | "*") ~ (ASCII_ALPHANUMERIC | "_" | "-" | ".")* ~ !"::" }
```
Comment (sinkingsugar) [inline shards/lang/src/shards.pest]: Can you just run a quick test on performance, time. Sometimes pegs fuck up

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

File: shards/lang/src/shards.pest (hunk 6 above)
Comment (guusw) [reply shards/lang/src/shards.pest]:
```
/a/tst>./script.sh
 == Testing new == 

real    0m0.053s
user    0m0.000s
sys     0m0.000s
 == Testing old ==

real    0m0.049s
user    0m0.000s
sys     0m0.000s
/a/tst>./script.sh
 == Testing old == 

real    0m0.051s
user    0m0.000s
sys     0m0.000s
 == Testing new ==

real    0m0.054s
user    0m0.000s
sys     0m0.000s
/a/tst>./script.sh
 == Testing old == 

real    0m0.050s
user    0m0.000s
sys     0m0.000s
 == Testing new ==

real    0m0.053s
user    0m0.000s
sys     0m0.000s
/a/tst>./script.sh
 == Testing old == 

real    0m0.049s
user    0m0.000s
sys     0m0.000s
 == Testing new ==

real    0m0.051s
user    0m0.000s
sys     0m0.000s
/a/tst>
```

Comment (guusw) [reply shards/lang/src/shards.pest]: Tested some ambal scripts, all with similar margins, the above one is `card-stats-library\ \(aMzdScjAE\).shs`

Comment (guusw) [reply shards/lang/src/shards.pest]: Running `shards format ./in.shs >/dev/null`

File: include/shards/shards.h (hunk 3 above)
Comment (guusw) [reply include/shards/shards.h:1046]: Doesn't use SHString

**Review by sinkingsugar — APPROVED** [review]

**Pull request merged**